src/NHibernate/Async/Action/CollectionRecreateAction.cs

```diff
@@ -41,9 +41,12 @@ public override async Task ExecuteAsync(CancellationToken cancellationToken)
 
 			await (PreRecreateAsync(cancellationToken)).ConfigureAwait(false);
 
-			await (Persister.RecreateAsync(collection, Key, Session, cancellationToken)).ConfigureAwait(false);
+			var key = await (GetKeyAsync(cancellationToken)).ConfigureAwait(false);
+			await (Persister.RecreateAsync(collection, key, Session, cancellationToken)).ConfigureAwait(false);
 
-			await (Session.PersistenceContext.GetCollectionEntry(collection).AfterActionAsync(collection, Session, cancellationToken)).ConfigureAwait(false);
+			var entry = Session.PersistenceContext.GetCollectionEntry(collection);
+			entry.CurrentKey = key;
```

Comment:
I think setting `CurrentKey` here is more correct than setting only `loadedKey` in `AfterAction`.